### PR TITLE
LoadBalancer Rule support

### LINKED ISSUE #72: Add support for Load Balancer Rule operations
Basic CRUD operations for Load Balancers is already available. But they are not
all that useful until you can also manage Load Balancer Rules.

Hence we need to add support for all the rule operations available in the API:
https://developers.katapult.io/api/docs/latest/groups/load_balancers/rules/

### PR DESCRIPTION
**Pull request by strideynet**
Resolves #72 and unblocks work on CCM

todo:
- [x] Tests

### CODE REVIEW

File: core/load_balancer_rule_test.go [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testJSONMarshaling(t, tt.obj)
+		})
+	}
+}
+
+func TestLoadBalancerRulesClient_Delete(t *testing.T) {
```
Comment (jimeh) [inline core/load_balancer_rule_test.go:101]: nitpick: I like placing the tests in the same order as the functions they're testing, just so the placement/position between the implementation and test files correlate 😁

File: core/load_balancer_rule.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	"net/url"
+
+	"github.com/krystal/go-katapult"
+)
+
+type LoadBalancerRuleAlgorithm string
+
+const (
+	RoundRobinRuleAlgorithm LoadBalancerRuleAlgorithm = "round_robin"
+	//nolint:lll
```
Comment (jimeh) [inline core/load_balancer_rule.go:15]: nitpick: I think ignoring the `lll` linter for this whole const block looks cleaner :)

```suggestion
//nolint:lll
const (
	RoundRobinRuleAlgorithm       LoadBalancerRuleAlgorithm = "round_robin"
```

**Review by jimeh — CHANGES_REQUESTED** [review]
This looks very good, I particularly like the `fakeRequestMaker` stcut to simplify test bodies :)

There's just a couple of missing tests, and a slight method signature inconsistency which we might want to discuss a bit further.

File: core/load_balancer_rule_test.go [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+			}},
+			wantResp: &katapult.Response{
+				Pagination: &katapult.Pagination{Total: 333},
+			},
+			listOptions: &ListOptions{
+				Page:    5,
+				PerPage: 32,
+			},
+			frm: fakeRequestMakerArgs{
+				wantPath:   "/core/v1/load_balancers/xyzzy/rules?page=5&per_page=32", //nolint:lll
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/load_balancer_rule_test.go:386]: That's a really nice idea 👍

File: core/load_balancer_rule_test.go [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+			name:           "success",
+			loadBalancerID: "xyzzy",
+			want: []LoadBalancerRule{{
+				ID:              "abc",
+				DestinationPort: 666,
+			}},
+			wantResp: &katapult.Response{
+				Pagination: &katapult.Pagination{Total: 333},
+			},
+			listOptions: &ListOptions{
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/load_balancer_rule_test.go:381]: Good shout 👍

File: core/load_balancer_rule_test.go [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+		name string
+		frm  fakeRequestMakerArgs
+
+		loadBalancerID string
+		listOptions    *ListOptions
+
+		want         []LoadBalancerRule
+		wantResp     *katapult.Response
+		wantErr      string
+		wantResponse bool
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/load_balancer_rule_test.go:369]: Which field is this referring to

[the remark this answers is not in the document]
Comment (jimeh) [reply core/load_balancer_rule_test.go:369]: Oh, `wantResponse`. Only unless I'm blind, only places it appears is in the struct definitions :)

File: core/client_test.go [hunk 6]
```diff
@@ -100,16 +101,50 @@ func TestClientImplementsRequestMaker(t *testing.T) {
...
 ) (*katapult.Response, error) {
-	return nil, nil
+	assert.True(frm.t, frm.newRequestBeenCalled)
+	assert.NotNil(frm.t, req)
+
+	if frm.args.doErr != nil {
+		return frm.args.doResp, frm.args.doErr
+	}
+
+	if frm.args.doResponseBody != nil {
```
[the remark this answers is not in the document]
Comment (strideynet) [reply core/client_test.go:136]: Hmm, I'll have a look at this.

**Pull request merged**